all_errors_are_fatal Configuration to Throw All Expected and Unexpected Exceptions

devstein opened the pull request:
Closes https://github.com/jupyter-incubator/sparkmagic/issues/578

Creates an `all_errors_are_fatal` configuration to throw any error when executing a notebook. This is useful for executing notebooks via `papermill` in a production setting where an error in any notebook cell should cause the notebook to fail with `exit code 1`

CC @itamarst @PedroRossi

itamarst commented:
This seems like it would fix the problem, yes, but the additional feature also sounds good if you want to add another flag with another PR.

Alternatively, insofar as `spark_statement_errors_are_fatal` never actually worked, I'd be happy to accept the `all_errors_are_fatal` PR instead of this one.

itamarst commented:
So next step: either fix failing test, or, change PR to replace it with more general feature (all errors cause processing to stop).

itamarst commented:
And, what I should've said first: thank you for the PR!

PedroRossi commented:
@devstein thanks for the `all_errors_are_fatal` catch, this is perfect for production environments using papermill (probably we are doing the same thing :smile: )

devstein commented:
@PedroRossi Of course! 

@itamarst Update to use a `all_errors_are_fatal` configuration. Do you mind helping with the import error in the `python 2.7.13` test?

itamarst commented on README.md:
````diff
@@ -81,7 +81,8 @@ If you want Papermill rendering to stop on a Spark error, edit the `~/.sparkmagi
 ```json
 {
     "spark_statement_errors_are_fatal": true,
````
I believe this particular config option was added specifically for the Papermill use case, but doesn't actually work—maybe it should be ripped out?

devstein commented:
@itamarst Good call, do you have a preference on how I removed the unnecessary `spark_errors_are_fatal` flag?


Option 1: Always raise `SparkStatementException` and handle as an expected exception

1. Removed `spark_errors_are_fatal` flag
2. Always raise a `SparkStatementException` on spark errors
3. By default, `SparkStatementException` is an expected exception and is displayed in the notebook as an error 
4. If all_errors_are_fatal flag is enabled, `SparkStatementException`, like all exceptions, will not be caught


Option 2: Only raise `SparkStatementException` on `all_errors_are_fatal`

1. Removed `spark_errors_are_fatal` flag
2. By default output spark errors do not raise an exception and are outputted to the notebook (`ipython_display.send_error(out)`) 
3. If `all_errors_are_fatal` flag is enabled, spark errors will raise a `SparkStatementException`, and like all exceptions, it will not be caught

I prefer option 1 because I don't like the idea of always swallowing spark errors and then raising a new error when an unrelated flag `all_errors_fatal` is set.  But I don't know how raising `SparkStatementException` will affect other parts of the codebase that don't `handle_expected_exceptions`

devstein commented:
@itamarst I updated the code and the tests to `Option 1` above. Let me know if you need me to make changes.

itamarst commented:
Thank you! Taking a look now.

itamarst commented:
@devstein 

Would you be interested in getting commit access to the repo and becoming a maintainer?

You need to:

1. Agree to the code of conduct.
2. Follow best practices (automated tests, manual test, docs, reviews) when adding code.

devstein commented:
@itamarst 

I'm definitely interested in becoming a maintainer.

Beyond the code of conduct and maintaining best practices, what are the time expectations? How are features and issues prioritized? Is there a feature roadmap or long-term priorities?

itamarst commented:
@devstein right now I just do a few hours a month, so it can be as little or as many as you want. Given limited time I focus on PRs from others, mostly.

I feel like long-term priority is maintainability, most important being testing which is far too manual at the moment. I opened some issues here: https://github.com/jupyter-incubator/sparkmagic/milestone/7

devstein commented:
@itamarst Okay, that makes sense. I'm happy to help
